### PR TITLE
[SPARK-32055][CORE][SQL] Unify getReader and getReaderForRange in ShuffleManager

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

This PR tries to unify the method `getReader` and `getReaderForRange` in `ShuffleManager`.

### Why are the changes needed?

Reduce the duplicate codes, simplify the implementation, and for better maintenance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Covered by existing tests.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 1]
```diff
@@ -322,7 +322,7 @@ private[spark] abstract class MapOutputTracker(conf: SparkConf) extends Logging
   // For testing
   def getMapSizesByExecutorId(shuffleId: Int, reduceId: Int)
       : Iterator[(BlockManagerId, Seq[(BlockId, Long, Int)])] = {
-    getMapSizesByExecutorId(shuffleId, reduceId, reduceId + 1)
+    getMapSizesByExecutorId(shuffleId, mapStatus => (0, mapStatus.length), reduceId, reduceId + 1)
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:325]: Since `mapStatus => (0, mapStatus.length)` seems to be used frequently, do you think we can define this function somewhere to prevent mistake?

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:325]: yes, sounds good.

**Comment by Ngone51** [comment]
@JkSelf @cloud-fan Could you please take a look? Thanks!

**Comment by Ngone51** [comment]
Jenkins, retest this please.

**Comment by Ngone51** [comment]
Jenkins, retest this please.

File: core/src/main/scala/org/apache/spark/shuffle/ShuffleManager.scala [hunk 2]
```diff
@@ -43,26 +44,16 @@ private[spark] trait ShuffleManager {
...
+   * endMapIndex-1, inclusive).
+   *
    * Called on executors by reduce tasks.
    */
-  def getReaderForRange[K, C](
+  def getReader[K, C](
       handle: ShuffleHandle,
-      startMapIndex: Int,
-      endMapIndex: Int,
+      mapIndexRange: Array[MapStatus] => (Int, Int),
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/shuffle/ShuffleManager.scala:56]: It's a function only because we need to know the total number of mappers? Maybe we can still use `startMapIndex: Int` and `endMapIndex: Int`, caller side can use `Int.Max` as `endMapIndex`, to indicate that it wants to read all the mappers.

Comment (JkSelf) [inline core/src/main/scala/org/apache/spark/shuffle/ShuffleManager.scala:56]: Same with the parameters of `startPartition` and `endPartition`, here we can remain the `startMapIndex` and `endMapIndex` in getReader method. It will be more clearly.

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleManager.scala:56]: The problem is we can not know the `endMapIndex` at the caller side. Maybe we could try @cloud-fan 's suggestion above.

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleManager.scala:56]: > It's a function only because we need to know the total number of mappers?

Yes.

Just a little worry that `Int.Max` is not easy to know from intuitive. It also bring small trouble for log info in `getMapSizesByExecutorId` since you can't know the real `endMapIndex` there.

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleManager.scala:56]: We can document it, it's an internal API anyway.

For logging, we do know the number of total mappers inside the method, right?

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleManager.scala:56]: Ok, I get it how to know the `endMapIndex`.

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 3]
```diff
@@ -335,28 +335,12 @@ private[spark] abstract class MapOutputTracker(conf: SparkConf) extends Logging
...
-   *         tuples describing the shuffle blocks that are stored at that block manager.
-   */
-  def getMapSizesByRange(
       shuffleId: Int,
       startMapIndex: Int,
       endMapIndex: Int,
       startPartition: Int,
-      endPartition: Int): Iterator[(BlockManagerId, Seq[(BlockId, Long, Int)])]
+      endPartition: Int)
+  : Iterator[(BlockManagerId, Seq[(BlockId, Long, Int)])]
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:343]: unnecessary change

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 4]
```diff
@@ -737,35 +721,21 @@ private[spark] class MapOutputTrackerMaster(
...
-  override def getMapSizesByRange(
       shuffleId: Int,
       startMapIndex: Int,
       endMapIndex: Int,
       startPartition: Int,
-      endPartition: Int): Iterator[(BlockManagerId, Seq[(BlockId, Long, Int)])] = {
-    logDebug(s"Fetching outputs for shuffle $shuffleId, mappers $startMapIndex-$endMapIndex" +
-      s"partitions $startPartition-$endPartition")
+      endPartition: Int)
+    : Iterator[(BlockManagerId, Seq[(BlockId, Long, Int)])] = {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:729]: unnecessary change

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 5]
```diff
@@ -737,35 +721,21 @@ private[spark] class MapOutputTrackerMaster(
...
+    logDebug(s"Fetching outputs for shuffle $shuffleId")
     shuffleStatuses.get(shuffleId) match {
-      case Some(shuffleStatus) =>
+      case Some (shuffleStatus) =>
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:732]: unnecessary change.

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 6]
```diff
@@ -737,35 +721,21 @@ private[spark] class MapOutputTrackerMaster(
...
         shuffleStatus.withMapStatuses { statuses =>
+          val endMapIndex0 = if (endMapIndex == Int.MaxValue) statuses.length else endMapIndex
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:734]: `actualEndMapIndex`?

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 7]
```diff
@@ -737,35 +721,21 @@ private[spark] class MapOutputTrackerMaster(
...
+          logDebug(s"Convert map statuses for shuffle $shuffleId, " +
+            s"partitions $startPartition-$endPartition, mappers $startMapIndex-$endMapIndex0")
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:736]: let's follow the original log and put `mappers` before `partitions`.

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 8]
```diff
@@ -800,35 +770,20 @@ private[spark] class MapOutputTrackerWorker(conf: SparkConf) extends MapOutputTr
...
-  override def getMapSizesByRange(
       shuffleId: Int,
       startMapIndex: Int,
       endMapIndex: Int,
       startPartition: Int,
-      endPartition: Int): Iterator[(BlockManagerId, Seq[(BlockId, Long, Int)])] = {
-    logDebug(s"Fetching outputs for shuffle $shuffleId, mappers $startMapIndex-$endMapIndex" +
-      s"partitions $startPartition-$endPartition")
+      endPartition: Int)
+    : Iterator[(BlockManagerId, Seq[(BlockId, Long, Int)])] = {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:778]: unnecessary change

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 9]
```diff
@@ -800,35 +770,20 @@ private[spark] class MapOutputTrackerWorker(conf: SparkConf) extends MapOutputTr
...
+    logDebug(s"Fetching outputs for shuffle $shuffleId")
     val statuses = getStatuses(shuffleId, conf)
     try {
+      val endMapIndex0 = if (endMapIndex == Int.MaxValue) statuses.length else endMapIndex
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:782]: ditto

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 10]
```diff
@@ -800,35 +770,20 @@ private[spark] class MapOutputTrackerWorker(conf: SparkConf) extends MapOutputTr
...
+      logDebug(s"Convert map statuses for shuffle $shuffleId, " +
+        s"partitions $startPartition-$endPartition, mappers $startMapIndex-$endMapIndex0")
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:784]: ditto

File: core/src/main/scala/org/apache/spark/shuffle/ShuffleManager.scala [hunk 11]
```diff
@@ -43,23 +44,15 @@ private[spark] trait ShuffleManager {
       context: TaskContext,
       metrics: ShuffleWriteMetricsReporter): ShuffleWriter[K, V]
 
-  /**
-   * Get a reader for a range of reduce partitions (startPartition to endPartition-1, inclusive).
-   * Called on executors by reduce tasks.
-   */
-  def getReader[K, C](
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/shuffle/ShuffleManager.scala:50]: can we keep this method as an overload? We can mark it as final and simply call the other `getReader` with `0, Int.Max` as the map index parameters.

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala (hunk 5 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:732]: Weird. It's actually the same with master:
https://github.com/apache/spark/blob/ef51ff9dc8c220fcbed76cdd1783f58f400df48c/core/src/main/scala/org/apache/spark/MapOutputTracker.scala#L746

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:732]: And also `: Iterator[(BlockManagerId, Seq[(BlockId, Long, Int)])] = {`

File: core/src/main/scala/org/apache/spark/shuffle/ShuffleManager.scala (hunk 11 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleManager.scala:50]: ok, sounds good!

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala (hunk 4 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:729]: this actually extends from `getMapSizesByExecutorId` above and its indent is wrong previously.

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala (hunk 8 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:778]: this actually extends from getMapSizesByExecutorId above and nothing changes.

**Comment by Ngone51** [comment]
Jenkins, retest this please.

**Comment by Ngone51** [comment]
Jenkins, retest this please.

**Comment by Ngone51** [comment]
Jenkins, retest this please.

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 12]
```diff
@@ -335,28 +335,12 @@ private[spark] abstract class MapOutputTracker(conf: SparkConf) extends Logging
    *         tuples describing the shuffle blocks that are stored at that block manager.
    */
   def getMapSizesByExecutorId(
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:337]: can we keep it as overload as well?

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala (hunk 5 above)
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:732]: I think `Some(shuffleStatus)` is the correct code style

File: core/src/main/scala/org/apache/spark/rdd/SubtractedRDD.scala [hunk 13]
```diff
@@ -23,13 +23,7 @@ import scala.collection.JavaConverters._
...
 import scala.reflect.ClassTag
 
-import org.apache.spark.Dependency
-import org.apache.spark.OneToOneDependency
-import org.apache.spark.Partition
-import org.apache.spark.Partitioner
-import org.apache.spark.ShuffleDependency
-import org.apache.spark.SparkEnv
-import org.apache.spark.TaskContext
+import org.apache.spark.{Dependency, MapOutputTracker, OneToOneDependency, Partition, Partitioner, ShuffleDependency, SparkEnv, TaskContext}
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/rdd/SubtractedRDD.scala:26]: unnecessary change

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala (hunk 5 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:732]: yea, make senses.

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala (hunk 12 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:337]: It would be only used by 3 tests yet. Do you think it's necessary?

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala (hunk 4 above)
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:729]: The previous indentation was not wrong... both are OK but it's better to save one line and keep the code diff smaller

**Comment by Ngone51** [comment]
retest this please.

**Review by cloud-fan — APPROVED** [review]

**Comment by Ngone51** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 14]
```diff
@@ -335,23 +335,6 @@ private[spark] abstract class MapOutputTracker(conf: SparkConf) extends Logging
...
-      startPartition: Int,
-      endPartition: Int)
-  : Iterator[(BlockManagerId, Seq[(BlockId, Long, Int)])]
-
-  /**
-   * Called from executors to get the server URIs and output sizes for each shuffle block that
-   * needs to be read from a given range of map output partitions (startPartition is included but
-   * endPartition is excluded from the range) and is produced by
-   * a range of mappers (startMapIndex, endMapIndex, startMapIndex is included and
-   * the endMapIndex is excluded).
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:348]: Hi, @Ngone51 . This should be the function description of the unified `getMapSizesByExecutorId`.

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:348]: Specifically, `endMapIndex == Int.MaxValue` seems worth to mention in a way here.

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:348]: Yea, updated, thanks!

**Comment by cloud-fan** [comment]
retest this please

**Review by cloud-fan — APPROVED** [review]

**Comment by Ngone51** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by dongjoon-hyun** [comment]
+1, late LGTM. Thank you, @Ngone51 and @cloud-fan .

**Comment by Ngone51** [comment]
thanks all!
